### PR TITLE
Add validation to select fields and add placeholder option to additional fields API

### LINKED ISSUE #49922: Required select fields added using Additional Checkout Field API do not have an initial value set
### Prerequisites

- [X] I have carried out troubleshooting steps and I believe I have found a bug.
- [X] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

When adding required select fields using Additional Checkout Field API, if the value goes unmodified (e.g. if the user is fine with the first selected option) then this option is not actually set on the data store. The field has an empty value when submitting the checkout form. This results in an error like so:

image

### Expected behavior

If a field is required, the value in the data store should be the field's first option.

### Actual behavior

If not modified, the field has an empty value when submitting the checkout form

### Steps to reproduce

1. Install this plugin - [additional-checkout-fields-tester.zip](https://github.com/user-attachments/files/16364487/additional-checkout-fields-tester.zip)
2. Go to the Checkout block and fill in your details. **Do not modify the select fields**.
3. Try to check out.
4. Observe the error shown after the checkout fails.

### WordPress Environment

n/a

### Isolating the problem

- [X] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [X] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/products/storefront/).
- [X] I can reproduce this bug consistently using the steps above.

### PR DESCRIPTION
**Pull request by senadir**
### Changes proposed in this Pull Request:

This is a follow up to https://github.com/woocommerce/woocommerce/pull/49616, here, I add validation to all fields, not just country and state, by abstracting the validation a bit higher.

This also tweaks additional fields API, instead of passing an empty option, the shopper can select the "placeholder" to indicate no option, this can be debated. When an additional fields is optional, that option can be selected, otherwise it's disabled like state and country.

I also added a "placeholder" prop to additional fields API, this allows developer to replace the placeholder if it's too wordy.

Closes https://github.com/woocommerce/woocommerce/issues/49922

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:
0. Install this updated version of [additional-checkout-fields-tester.zip](https://github.com/user-attachments/files/16365659/additional-checkout-fields-tester.zip)
1. In Checkout, don't fill out anything.
2. Try placing an order, notice that "What is your main role at your company?" field is showing an error.
3. Notice that "Preferred time to contact (Morning, Afternoon, or Evening)" is showing an error.
4. Notice that "Which neighbor should we leave it with if unable to hide?" is not showing an error.
5. Notice that "What is your main role at your company?" value is "Select a role".
6. Notice that "Preferred time to contact (Morning, Afternoon, or Evening)" value is "Select a time"
7. Notice that "Which neighbor should we leave it with if unable to hide?" value is "Select a Which neighbor should we leave it with if unable to hide? (optional)".
8. Selecting values for the first select should clear the error.
9. For the Select a role, make sure you can't select it back.
10. Run the testing steps in this PR https://github.com/woocommerce/woocommerce/pull/49616 to make sure we didn't regress it when we moved the code.

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

-   [x] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [x] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

Add placeholder options and validation to all checkout select inputs.

</details>

### CODE REVIEW

File: plugins/woocommerce-blocks/assets/js/base/components/cart-checkout/form/form.tsx [hunk 1]
```diff
@@ -292,6 +292,10 @@ const Form = < T extends AddressFormValues | ContactFormValues >( {
 								} );
 							} }
 							options={ field.options }
+							required={ field.required }
+							errorMessage={
+								fieldProps.errorMessage || undefined
+							}
```
Comment (senadir) [inline plugins/woocommerce-blocks/assets/js/base/components/cart-checkout/form/form.tsx:304]: Not sure those are needed but the linter kept screaming at me.

File: plugins/woocommerce-blocks/assets/js/base/components/country-input/country-input.tsx [hunk 2]
```diff
@@ -25,58 +21,27 @@ export const CountryInput = ( {
...
 	}, [ countries ] );
 
-	const validationError = useSelect( ( select ) => {
-		const store = select( VALIDATION_STORE_KEY );
-		return (
-			store.getValidationError( errorId || '' ) || {
-				hidden: true,
-			}
-		);
-	} );
```
Comment (senadir) [inline plugins/woocommerce-blocks/assets/js/base/components/country-input/country-input.tsx:57]: This all moved to Select.

File: plugins/woocommerce-blocks/assets/js/base/components/country-input/country-input.tsx [hunk 3]
```diff
@@ -25,58 +21,27 @@ export const CountryInput = ( {
...
-
 	return (
-		<div
-			className={ clsx( className, 'wc-block-components-country-input', {
-				'has-error': ! validationError.hidden,
-			} ) }
-		>
```
Comment (senadir) [inline plugins/woocommerce-blocks/assets/js/base/components/country-input/country-input.tsx:64]: This is now in Select.

File: plugins/woocommerce-blocks/assets/js/base/components/select/index.tsx [hunk 4]
```diff
@@ -22,61 +27,151 @@ type SelectProps = Omit<
...
+		...restOfProps
+	} = props;
 	const selectOnChange = useCallback(
 		( event: React.ChangeEvent< HTMLSelectElement > ) => {
 			onChange( event.target.value );
 		},
 		[ onChange ]
 	);
 
+	const emptyOption: SelectOption = useMemo(
```
Comment (senadir) [inline plugins/woocommerce-blocks/assets/js/base/components/select/index.tsx:56]: This is now a `useMemoe` because it was missing from the `options` useMemo.

File: plugins/woocommerce-blocks/assets/js/base/components/select/index.tsx [hunk 5]
```diff
@@ -22,61 +27,151 @@ type SelectProps = Omit<
...
+			clearValidationError( errorIdString );
+		};
+	}, [
+		clearValidationError,
+		value,
+		errorIdString,
+		errorMessage,
+		required,
+		setValidationErrors,
+	] );
```
Comment (senadir) [inline plugins/woocommerce-blocks/assets/js/base/components/select/index.tsx]: Copy pasted from combobox.

File: plugins/woocommerce-blocks/assets/js/base/components/select/index.tsx [hunk 6]
```diff
@@ -22,61 +27,151 @@ type SelectProps = Omit<
...
+						className="wc-blocks-components-select__select"
+						id={ inputId }
+						size={ size !== undefined ? size : 1 }
+						onChange={ selectOnChange }
+						value={ value }
+						aria-invalid={
+							error?.message && ! error?.hidden ? true : false
+						}
+						aria-errormessage={ validationErrorId }
+						{ ...restOfProps }
```
Comment (senadir) [inline plugins/woocommerce-blocks/assets/js/base/components/select/index.tsx:150]: also from combobox.

File: plugins/woocommerce-blocks/assets/js/base/components/select/index.tsx [hunk 7]
```diff
@@ -22,61 +27,151 @@ type SelectProps = Omit<
...
+						className="wc-blocks-components-select__expand"
+						icon={ chevronDown }
+					/>
+				</div>
 			</div>
+			{ validationError && validationError.hidden !== true && (
+				<ValidationInputError
+					errorMessage={ validationError.message }
+				/>
+			) }
```
Comment (senadir) [inline plugins/woocommerce-blocks/assets/js/base/components/select/index.tsx]: I could have done

```jsx
	<ValidationInputError propertyName={ errorId } />
```

But I didn't to match the original select.

File: plugins/woocommerce-blocks/packages/components/checkbox-control/index.tsx [hunk 8]
```diff
@@ -32,6 +32,8 @@ export const CheckboxControl = ( {
 	hasError = false,
 	checked = false,
 	disabled = false,
+	errorId,
+	errorMessage,
```
Comment (senadir) [inline plugins/woocommerce-blocks/packages/components/checkbox-control/index.tsx:36]: was showing an error in console.

File: plugins/woocommerce-blocks/assets/js/base/components/select/index.tsx [hunk 9]
```diff
@@ -22,61 +27,151 @@ type SelectProps = Omit<
...
 		[ onChange ]
 	);
 
+	const emptyOption: SelectOption = useMemo(
+		() => ( {
+			value: '',
+			label:
+				placeholder ??
+				sprintf(
+					// translators: %s will be label of the country input. For example "country/region".
```
Comment (samueljseay) [inline plugins/woocommerce-blocks/assets/js/base/components/select/index.tsx]: This translator comment is no longer correct

File: plugins/woocommerce-blocks/assets/js/base/components/select/index.tsx (hunk 7 above)
Comment (samueljseay) [reply plugins/woocommerce-blocks/assets/js/base/components/select/index.tsx]: @senadir why is it so important to match the original? We know there are a few bugs and things done incorrectly lying around.

File: plugins/woocommerce-blocks/assets/js/base/components/state-input/state-input.tsx [hunk 10]
```diff
@@ -117,35 +88,23 @@ const StateInput = ( {
...
-					<ValidationInputError
-						errorMessage={ validationError.message }
-					/>
-				) }
-			</div>
+				options={ options }
+				label={ label || '' }
+				id={ id }
+				onChange={ ( newValue ) => {
+					if ( required ) {
```
Comment (samueljseay) [inline plugins/woocommerce-blocks/assets/js/base/components/state-input/state-input.tsx]: looks like some code I left in by accident? this can be removed

File: plugins/woocommerce-blocks/assets/js/base/components/select/index.tsx (hunk 6 above)
Comment (samueljseay) [reply plugins/woocommerce-blocks/assets/js/base/components/select/index.tsx:150]: Is there a reason for just copy pasting `...restOfProps`. I'm not sure I understand the approach of just copy/pasting everything from the old component. If something doesn't work perhaps we should figure out what props are not being passed and just explicitly pass them.

There is a bit of a tendency higher up in the component tree to pass down all props and you get react warnings because some are not props that are legal on the html component.

File: plugins/woocommerce-blocks/assets/js/base/components/select/index.tsx [hunk 11]
```diff
@@ -22,61 +27,151 @@ type SelectProps = Omit<
...
+			disabled: !! required,
+		} ),
+		[ label, placeholder, required ]
+	);
+	const errorIdString = errorId || restOfProps.id || '';
+	const optionsWithEmpty = useMemo< SelectOption[] >( () => {
+		return [ emptyOption ].concat( options );
+	}, [ options, emptyOption ] );
+
+	const { setValidationErrors, clearValidationError } =
```
Comment (samueljseay) [inline plugins/woocommerce-blocks/assets/js/base/components/select/index.tsx:83]: Overall i do think the validation being here is better. It'd be nice if we could opt out of the placeholder option but we don't actually have a need for that right now do we?

File: plugins/woocommerce-blocks/packages/components/checkbox-control/index.tsx (hunk 8 above)
Comment (samueljseay) [reply plugins/woocommerce-blocks/packages/components/checkbox-control/index.tsx:36]: I think it would be more correct to ensure we don't pass invalid props to this component thats why the warning happens.

**Review by samueljseay — APPROVED** [review]
A few nitpicks from me, but it tests well and overall I think the approaches are good 🚢

File: plugins/woocommerce-blocks/assets/js/base/components/select/index.tsx (hunk 9 above)
Comment (opr) [inline plugins/woocommerce-blocks/assets/js/base/components/select/index.tsx]:
```suggestion
					// translators: %s will be label of the field. For example "country/region".
```

File: plugins/woocommerce-blocks/assets/js/base/components/select/index.tsx (hunk 6 above)
Comment (senadir) [reply plugins/woocommerce-blocks/assets/js/base/components/select/index.tsx:150]: `...restOfProps` was already there in the select so it's not new addition, I only brought back `aria-invalid` and `aria-errormessage`, those are not supported by the old combobox but supported by the new select.

File: plugins/woocommerce-blocks/assets/js/base/components/select/index.tsx (hunk 7 above)
Comment (senadir) [reply plugins/woocommerce-blocks/assets/js/base/components/select/index.tsx]: When I say original Select I meant the select component you used in country and state.

For the combobox, it uses `ValidationInputError` with `propertyName`, the new country and state select, it uses the above. I didn't want to modify your code because I wanted to keep this a light refactor.

File: plugins/woocommerce-blocks/packages/components/checkbox-control/index.tsx (hunk 8 above)
Comment (senadir) [reply plugins/woocommerce-blocks/packages/components/checkbox-control/index.tsx:36]: Yeah I will strip them out in Form.

File: plugins/woocommerce-blocks/assets/js/base/components/select/index.tsx (hunk 11 above)
Comment (senadir) [reply plugins/woocommerce-blocks/assets/js/base/components/select/index.tsx:83]: Yes I can't think of a use case yet unless we start having default values but those should be set server-side.

File: plugins/woocommerce-blocks/docs/third-party-developers/extensibility/checkout-block/additional-checkout-fields.md [hunk 12]
```diff
@@ -292,12 +292,15 @@ As well as the options above, text fields also support a `required` option. If t
...
+
+You can set a placeholder to be shown on the select by passing a `placeholder` value when registering the field. This will be the first option in the select and will not be selectable if the field is required.
 
 | Option name | Description | Required? | Example        | Default value |
 |-----|-----|-----|----------------|--------------|
 | `options` | An array of options to show in the select input. Each options must be an array containing a `label` and `value` property. Each entry must have a unique `value`. Any duplicate options will be removed. The `value` is what gets submitted to the server during checkout and the `label` is simply a user-friendly representation of this value. It is not transmitted to the server in any way. | Yes | see below | No default - this must be provided. |
 | `required` | If this is `true` then the shopper _must_ provide a value for this field during the checkout process. | No | `true` | `false` |
+| `placeholder` | If this value is set, the shopper will see this option in select. If the select is required, the shopper cannot select this option. | No | `Select a role | Select a $label |
```
Comment (opr) [inline plugins/woocommerce-blocks/docs/third-party-developers/extensibility/checkout-block/additional-checkout-fields.md]:
```suggestion
| `placeholder` | If this value is set, the shopper will see this option in the select. If the select is required, the shopper cannot select this option. | No | `Select a role | Select a $label |
```

File: plugins/woocommerce-blocks/docs/third-party-developers/extensibility/checkout-block/additional-checkout-fields.md [hunk 13]
```diff
@@ -425,11 +428,12 @@ add_action(
...
 		woocommerce_register_additional_checkout_field(
 			array(
-				'id'       => 'namespace/how-did-you-hear-about-us',
-				'label'    => 'How did you hear about us?',
-				'location' => 'order',
-				'type'     => 'select',
-				'options'  => [
+				'id'          => 'namespace/how-did-you-hear-about-us',
+				'label'       => 'How did you hear about us?',
+				'placeholder' => 'Select a source',,
```
Comment (opr) [inline plugins/woocommerce-blocks/docs/third-party-developers/extensibility/checkout-block/additional-checkout-fields.md]:
```suggestion
				'placeholder' => 'Select a source',
```

File: plugins/woocommerce-blocks/assets/js/base/components/select/index.tsx (hunk 11 above)
Comment (senadir) [reply plugins/woocommerce-blocks/assets/js/base/components/select/index.tsx:83]: Added an option to opt out, if a field is required and has a value, we skip the empty one.

File: plugins/woocommerce-blocks/assets/js/base/components/select/index.tsx (hunk 7 above)
Comment (senadir) [reply plugins/woocommerce-blocks/assets/js/base/components/select/index.tsx]: Updated.

Comment (senadir) [reply plugins/woocommerce-blocks/assets/js/base/components/select/index.tsx]: @samueljseay do you recall why you did:

```jsx
{ validationError && validationError.hidden !== true && (
	<ValidationInputError
		errorMessage={ validationError.message }
	/>
) }
```

instead of:

```jsx
<ValidationInputError propertyName={ errorId } />
```

File: plugins/woocommerce-blocks/assets/js/base/components/select/index.tsx (hunk 11 above)
Comment (opr) [reply plugins/woocommerce-blocks/assets/js/base/components/select/index.tsx:83]: > Added an option to opt out, if a field is required and has a value, we skip the empty one.

Is this possible to do from an extension's point of view?

**Review by opr — COMMENTED** [review]
Functionality wise this is working well for me, code looks good, besides the items Sam already identified. 

Tested this out and also tested with Store API.

File: plugins/woocommerce-blocks/assets/js/base/components/select/index.tsx (hunk 7 above)
Comment (samueljseay) [reply plugins/woocommerce-blocks/assets/js/base/components/select/index.tsx]: @senadir oh whoops that was a mistake on my part when I was using `ValidationInputError` I passed in errorMessage (I was having trouble getting the select to find the validation message for some reason) and then I didn't refactor this after I got that working. I realize now that is handled internally. The API of that component is a little confusing.

Comment (senadir) [reply plugins/woocommerce-blocks/assets/js/base/components/select/index.tsx]: It is confusing indeed, you're the second one getting confused by it this week.

File: plugins/woocommerce-blocks/assets/js/base/components/select/index.tsx (hunk 11 above)
Comment (senadir) [reply plugins/woocommerce-blocks/assets/js/base/components/select/index.tsx:83]: In what sense? Additional fields API? or someone using the select component?

They can simply pass a value other than `""`

Comment (opr) [reply plugins/woocommerce-blocks/assets/js/base/components/select/index.tsx:83]: When passing empty option it still renders the placeholder and an empty option as well. I don't think this is a problem.

Comment (senadir) [reply plugins/woocommerce-blocks/assets/js/base/components/select/index.tsx:83]: Yeah this is up to the consumer, I think we can filter out empty options if we want. For additional fields API empty option is no longer passed.
